README.md
# Бот-парсер интернет магазина Glam Go Gans

#### Описание

Этот бот был создан с целью практики таких технологий как:
- SQL (SQLite3)
- asyncio
- aiogram
- logging
- BeautifulSoup 4
- Selenium


history.txt
17.09.23:
    -Добавлена функция pars_url_artist (KASSAS20)
    -Добавлена функция pars_from_card_product(Firty28)
    -Добавлена функция get_soup(KASSAS20)
    -Добавлена функция get_soup_from_selenium(KASSAS20)
    -Добавлена функция pars_url_product(KASSAS20)
18.09.23:
    -Пофикшена функция pars_url_product(KASSAS20)
    -Добавлена функция pars_url_img(KASSAS20)
19.09.23:
    -Добавлена функционал комманды start в бота(KASSAS20)
    -Изменен формат входных данных функции pars_url_product(KASSAS20)
20.09.23:
    -Добавлен корректный вывод кнопок товаров в бота(KASSAS20)
    -Добавлена кнопка Домой в бот(KASSAS20)
    -Добавлен прототип функции не цензурной лексики в боте(KASSAS20)
    -Добавлена обработка неверных команд в боте(KASSAS20)
    -Добавлена функция вывода информации о продукте в боте(KASSAS20)
21.09.23:
    -Добавлена база данных(KASSAS20)
    -Добавлен конект бота с бд(KASSAS20)
    -Добавлен вывод картинок через url(KASSAS20)
    -Добавлено обновление бд раз в 6 часов(KASSAS20)
31.10.23:
    -Добавлена venv
    -Добавлен requirements.txt


test.py
e = {'CAKEBOY': {'ФУТБОЛКА «SMILES»': 'https://static.tildacdn.com/tild3932-6164-4736-a538-363338666433/4.jpg', 'ФУТБОЛКА «HAIRSTYLE»': 'https://static.tildacdn.com/tild3236-6633-4737-b438-396265663062/photo.jpg', 'ХУДИ «ОКОЛЬНЫЙ ПУТЬ НА ЗАПАД»':
                 'https://static.tildacdn.com/tild3938-3366-4136-a436-343436313163/OP4i35tbaJo.jpg'}, 'IROH': {'ФУТБОЛКА «XANNY»': 'https://static.tildacdn.com/tild3235-3539-4632-a539-643238623233/photo_2020-10-28_00-.jpg'}, 'Flipper Floyd': {None: None}}

url = 'https://glamgo.store/iroh/tproduct/237840861-165835212961-futbolka-xanny'

import pars_modul
from bs4 import BeautifulSoup as bs
html = pars_modul.get_soup_from_selenium(url)

with open('product.html', 'w') as file:
    file.write(str(html))
    file.close()

bot.py
import asyncio
import logging
from aiogram import Bot, Dispatcher, types
from aiogram.filters import Command
import config
import sqlite3


connect = sqlite3.connect('data.db')
cursor = connect.cursor()

cursor.execute('''SELECT * FROM parsing_data''')
data = cursor.fetchall()

artist_dict = {}
data_dict = {}

for i in data:
    if i[0] not in artist_dict:
        artist_dict[i[0]] = i[1]

    data_dict[i[2]] = [i[0], i[3], i[5], i[6], i[7], i[8], i[9], i[10], i[4]]


# Включаем логирование, чтобы не пропустить важные сообщения
logging.basicConfig(level=logging.INFO)

bot = Bot(token=config.TOKEN)
dp = Dispatcher()


@dp.message(Command("start"))
async def cmd_start(message: types.Message):
    kb = []
    for artist in artist_dict:
        kb.append([types.KeyboardButton(text=artist)])
    keyboard = types.ReplyKeyboardMarkup(keyboard=kb, resize_keyboard=True)
    await message.answer("Выбирите артиста", reply_markup=keyboard)


@dp.message()
async def button(message: types.Message):
    kb = []
    if message.text in artist_dict:  # проверяем есть ли введёный артист в словаре
        for position in data_dict:  # перебираем все товары
            print(position)
            if position != None and data_dict[position][0] == message.text:
                # создаём кнопку с именем товара
                kb.append([types.KeyboardButton(text=position)])
        kb.append([types.KeyboardButton(text='Домой')])
        keyboard = types.ReplyKeyboardMarkup(keyboard=kb, resize_keyboard=True)
        await message.answer(f'Выбери шмот', reply_markup=keyboard)

    elif message.text == 'Домой':
        await cmd_start(message)

    elif message.text in data_dict:
        for product in data_dict:
            if product == message.text:
                price = data_dict[message.text][2]
                type_product = data_dict[message.text][3]
                color = data_dict[message.text][4]
                structur = data_dict[message.text][5]
                application_chest = data_dict[message.text][6]
                application_back = data_dict[message.text][7]
                img = data_dict[message.text][8]

                mes = f'Товар: <a href="{img}" style="display:none;">{product}</a>\nЦена: {price}руб.\n{type_product}\n{color}\n{structur}\n{application_chest}\n{application_back}'
                while 'None' in mes:
                    mes = mes.replace('None', '')
                await message.answer(mes, parse_mode='HTML')
    else:
        await message.answer('Ты говоришь на непонятном мне языке')


async def main():
    await dp.start_polling(bot)

if __name__ == "__main__":
    asyncio.run(main())


data_base.py
import sqlite3
import pars_modul as pars
import os

file_path = "data.db"
try:
    os.remove(file_path)
    print(f"Файл {file_path} успешно удален.")
except FileNotFoundError:
    print(f"Файл {file_path} не найден.")
except Exception as e:
    print(f"Произошла ошибка при удалении файла: {e}")


connect_data = sqlite3.connect("data.db")
data = connect_data.cursor()

data.execute('''CREATE TABLE IF NOT EXISTS parsing_data(
                artist TEXT,
                url_artist TEXT,
                product TEXT,
                product_url TEXT,
                product_img TEXT,
                price TEXT,
                types TEXT,
                color TEXT,
                structure TEXT,
                application_chest TEXT,
                application_back TEXT)''')

url_artist_dict = pars.pars_url_arists()
url_product_dict = pars.pars_url_product(url_artist_dict)
img_dict = pars.pars_url_img(url_product_dict)

for artist in url_artist_dict:
    url_artist = url_artist_dict[artist]  # url_artist
    for product in url_product_dict[artist]:  # product
        url_product = url_product_dict[artist][product]  # url product
        if url_product != None:
            info_product = pars.pars_from_card_product(
                url_product)  # info product
            price = str(info_product['price'])
            types = str(info_product['types'])
            color = str(info_product['color'])
            structure = str(info_product['structure'])
            application_chest = str(info_product['application_chest'])
            application_back = str(info_product['application_back'])
            url_img = img_dict[artist][product]  # img
        else:
            price = None
            types = None
            color = None
            structure = None
            application_chest = None
            application_back = None
            url_img = None

        data.execute('''INSERT INTO parsing_data VALUES(?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)''', (artist, url_artist,
                     product, url_product, url_img, price, types, color, structure, application_chest, application_back))
        connect_data.commit()

connect_data.close()


pars_modul.py
from bs4 import BeautifulSoup as bs
import requests
from selenium import webdriver
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


def get_soup(url):  # получение html страницы через requests
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36'}
    response = requests.get(url, headers=headers)
    soup = bs(response.text, 'lxml')
    return soup


def get_soup_from_selenium(url):  # получение html страницы через селениум
    user_agent = "'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/75.0.3770.142 Safari/537.36'"
    option = webdriver.FirefoxOptions()
    option.add_argument("--headless")
    option.add_argument(f"user-agent={user_agent}")
    driver = webdriver.Firefox(options=option)
    wait = WebDriverWait(driver, 10)
    driver.get(url)
    wait.until(EC.presence_of_element_located(
        ('css selector', '#nav238128853 > div.t830__panel.t830__panel_bg.t830__panel_hover.t830__panel_open > div.t830__menu__content > div.t830__burger.t830__burger_mobile')))
    html = driver.page_source
    soup = bs(html, 'lxml')
    driver.quit()
    return soup


# извлекает ссылки на артистов с главной страницы
def pars_url_arists(soup=get_soup("https://glamgo.store/")):
    result = {}
    html_list_info_artist = soup.find(
        'li', class_='t229__list_item').find_all('li', class_="t-menusub__list-item t-name t-name_xs")
    for html_artist in html_list_info_artist:
        name_artist = html_artist.text.strip()
        url_card_artist = html_artist.find(
            'a', class_="t-menusub__link-item t-name t-name_xs").get('href').strip()
        result[name_artist] = url_card_artist
    return result


# сбор наименований и ссылок на товары с карточки артиста
def pars_url_product(dict_artist_url):
    dict_position = {}
    for artist in dict_artist_url:
        url = f'https://glamgo.store{dict_artist_url[artist]}'
        soup = get_soup_from_selenium(url)
        html_position = soup.find_all(
            'div', class_="js-product t-store__card t-col t-col_4 t-align_center t-item")
        dict_position[artist] = {}
        for position in html_position:

            name = position.find(
                'div', class_="js-store-prod-name js-product-name t-store__card__title t-typography__title t-name t-name_md").text.strip()
            href = position.find('a').get('href')
            dict_position[artist][name] = href
        if html_position == []:
            dict_position[artist] = {None: None}

    return dict_position


def pars_from_card_product(url):  # сбор информации со страницы товара
    soup = get_soup(url)
    name = soup.find_all('div', class_='t-store__prod-popup__title-wrapper')
    price = soup.find_all(
        'div', class_='js-product-price js-store-prod-price-val t-store__prod-popup__price-value')
    chars = soup.find_all('p', class_='js-store-prod-charcs')

    for i in name:
        name = i.text.strip()

    for i in price:
        price = i.text.strip()
    types = chars[0].text
    color = chars[1].text
    structure = chars[2].text
    try:
        application_chest = chars[3].text
    except Exception as _ex:
        application_chest = None
    try:
        application_back = chars[4].text
    except Exception as _ex:
        application_back = None

    chars_name = {
        'name': name,
        'price': price,
        'types': types,
        'color': color,
        'structure': structure,
        'application_chest': application_chest,
        'application_back': application_back
    }

    return chars_name


def pars_url_img(dict_position):  # парс ссылки на главную картинку карточек товаров
    dict_img = {}
    for artist in dict_position:
        dict_img[artist] = {}
        for product in dict_position[artist]:
            if product != None:
                soup = get_soup_from_selenium(dict_position[artist][product])
                try:
                    href = soup.find(
                        'div', class_='t-slds__bgimg t-bgimg js-product-img loaded').get('data-original')
                except AttributeError:
                    href = soup.find(
                        'meta', itemprop='image').get('content')
                dict_img[artist][product] = href
            else:
                dict_img[artist] = {None: None}

    return dict_img


main.py
from bs4 import BeautifulSoup as bs
import requests
import pars_modul

# soup = pars_modul.get_soup('https://glamgo.store/')
# dict_artist_url = pars_modul.pars_url_arists(soup)
# dict_position = pars_modul.pars_url_product(dict_artist_url)

dict_position = {'CAKEBOY': {'ФУТБОЛКА «SMILES»': 'https://glamgo.store/cakeboy/tproduct/237841683-283757633041-futbolka-smiles', 'ФУТБОЛКА «HAIRSTYLE»': 'https://glamgo.store/cakeboy/tproduct/237841683-354370475801-futbolka-hairstyle', 'ХУДИ «ОКОЛЬНЫЙ ПУТЬ НА ЗАПАД»': 'https://glamgo.store/cakeboy/tproduct/237841683-694602504431-hudi-okolnii-put-na-zapad'}, 'IROH': {'ФУТБОЛКА «XANNY»': 'https://glamgo.store/iroh/tproduct/237840861-165835212961-futbolka-xanny'}, 'Flipper Floyd': {None: None}}

dict_img = pars_modul.pars_url_img(dict_position)
print(dict_img)